rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+one C++ library that translates the Arrow RecordBatch to Tensors, which can also
+be reused in a TF op (as the downstream of the Dataset, or in a Python wrapper).
+
+# Alternatives Considered
+
+We’ve considered an alternative where
+[**StructuredTensor**](https://github.com/tensorflow/community/blob/master/rfcs/20190910-struct-tensor.md)
+is the unified in-memory representation, and **tf.Data** is the unified I/O
+abstraction.
+
+StructuredTensor is equally powerful as Arrow’s StructArray so it’s able to
+represent all our logical representations.
+
+Because StructuredTensor is a CompositeTensor, we could imagine that I/O +
+parsing of a logical data format in a physical storage format is handled by a
+specific tf.Data Dataset that yields StructuredTensors.
+
+We also need to be able to construct a beam source from such a Dataset (although
+there is a non-trivial gap).
+
+The advantages of this approach are:
+
+*   Less effort to integrate: adding a new format == adding a new tf.Dataset.
+*   Unified story across TF and beam.
+*   No third-party dependencies.
+    *   Note that we can still provide an Arrow -> StructuredTensor adapter to
+        achieve interoperability.
+
+What this alternative does not change / address:
+
+*   We still need the TensorAdapter API to convert from StructuredTensor to
+    other TF feedables, unless we are willing to only offer StructuredTensors to
+    our end-users (which might be the case eventually, but not likely to happen
+    soon). So a good portion of this proposal will remain mostly unchanged.
+*   We still cannot expose file-based or record-based interface to end-users in
+    the TF trainer. As that is a direct result of I/O + parsing being abstracted
+    out.
+
+The disadvantages of this approach are:
+
+*   This tightly couples TF with TFX components, in the following ways:
+
+    *   Components will need TF to read the data in.
+    *   Components that analyze the data (e.g. TFDV) will operate against
+        StructuredTensors, and the easiest way to to conduct certain
+        computations (for example, slicing a StructuredTensor or computing the
+        mean) with StructuredTensors is through TF ops and their python
+        bindings.
+
+    Such a coupling is not in our favor because:
+
+    *   Some TFX components functionally do not require TF to work. For example,
+        TFDV can analyze any data set. TFMA can analyze a model while treating
+        the model as a blackbox. In both cases, the ML library that trained the
+        model is irrelevant and TF should not be assumed. And coupling with TF
+        does not only introduce a heavy dependency, but also forces the user to
+        learn about TF if they need to implement a TFXIO for their data format.
+
+    *   Operations against StructuredTensors boils down to Python + TF ops. The
+        overhead of either is much higher than just calling an Arrow C++ API
+        that does the same operation.
```

Comment:
There has been interest in moving composite tensors down to C++, which would reduce this overhead. @jaingaurav  FYI

Reply:
Ack. We could revisit this point once there's a design for that.